### PR TITLE
[#3493] Add total attribute

### LINKED ISSUE #3493: Rename functional into total
The frontend currently transforms K's `total` attribute into Kore's `functional` attribute. We should also look into renaming our internal attribute to `Total` and start parsing `total` instead of `functional` (will need to be accompanied with a change to the frontend).

### PR DESCRIPTION
**Pull request by alexd1971**
Problem: Attribute `functional` is deprecated in favour of `total` and `K`-frontend currently rewrites `total` with `functional`. We need haskell-backend to support `total` attribute along with `functional` attribute for backward compatibility.

Solution: Add `total` attribute and make it interchangable with `functional` attribute.

Fixes #3493

### CODE REVIEW

**Review by ana-pantilie — CHANGES_REQUESTED** [review]
The total attribute should be used wherever the functional attribute is. I would suggest, instead of creating a new attribute type just modify the old functional type, rename it to total (and in all places we say functional say total instead) and make that attribute be parsed by either "functional" or "total".

**Comment by amelieled** [comment]
And don't forget to update the file `src/main/kore)/prelude.kore`! :)

File: kore/src/Kore/Attribute/Pattern/Total.hs [hunk 1]
```diff
@@ -0,0 +1,143 @@
+{- |
+Copyright   : (c) Runtime Verification, 2019-2021
```
Comment (ana-pantilie) [inline kore/src/Kore/Attribute/Pattern/Total.hs:2]:
```suggestion
Copyright   : (c) Runtime Verification, 2019-2023
```

File: kore/src/Kore/Attribute/Total.hs [hunk 2]
```diff
@@ -0,0 +1,69 @@
+{- |
+Module      : Kore.Attribute.Total
+Description : Total symbol attribute
+Copyright   : (c) Runtime Verification, 2018-2021
+License     : BSD-3-Clause
+Maintainer  : thomas.tuegel@runtimeverification.com
+-}
```
Comment (ana-pantilie) [inline kore/src/Kore/Attribute/Total.hs:7]:
```suggestion
{- |
Copyright   : (c) Runtime Verification, 2018-2023
License     : BSD-3-Clause
-}
```

File: kore/src/Kore/Internal/Key.hs [hunk 3]
```diff
@@ -54,7 +54,7 @@ import Prelude.Kore
 
 {- | A type for 'Key' attributes. We only need to keep track of the sort.
  Keys are constructor-like, and therefore they are always
- fully simplified, defined, and functional ML patterns,
+ fully simplified, defined, and total ML patterns,
```
Comment (ana-pantilie) [inline kore/src/Kore/Internal/Key.hs:57]:
```suggestion
 fully simplified, defined, and total function-like ML patterns,
```

File: kore/src/Kore/Internal/TermLike/TermLike.hs [hunk 4]
```diff
@@ -724,13 +724,13 @@ instance (Unparse variable, Ord variable) => Unparse (TermLike variable) where
...
-                        addFunctionalRepresentation $
+                        addTotalRepresentation $
                             addFunctionRepresentation $
                                 addDefinedRepresentation $
                                     addSimplifiedRepresentation $
                                         addConstructorLikeRepresentation []
-                addFunctionalRepresentation
-                    | Attribute.isFunctional $ termFunctional attrs = ("Fl" :)
+                addTotalRepresentation
+                    | Attribute.isTotal $ termTotal attrs = ("Fl" :)
```
Comment (ana-pantilie) [inline kore/src/Kore/Internal/TermLike/TermLike.hs:733]:
```suggestion
                    | Attribute.isTotal $ termTotal attrs = ("T" :)
```

You might need to regenerate some integration test output to add this change.

**Review by ana-pantilie — APPROVED** [review]
Looks great, thank you for making all the necessary changes! I have a few comments, please take a look at those and feel free to add `automerge` to it when you've resolved them.

File: kore/src/Kore/Internal/TermLike/TermLike.hs (hunk 4 above)
Comment (alexd1971) [reply kore/src/Kore/Internal/TermLike/TermLike.hs:733]: I'm going to do this in separate PR

**Pull request merged**